target/linux/rtl838x/files-5.4/drivers/gpio/gpio-rtl838x.c

```diff
@@ -0,0 +1,808 @@
+
+	v = rtl838x_r32(RTL838X_GPIO_PABC_DATA);
+	if (v & (1 << gpios->smi_data))
+		return 1;
+	return 0;
+}
+
+/* Tri-state of MDIO line */
+static void smi_z(struct rtl838x_gpios *gpios)
+{
+	/* MDIO pin to input */
+	rtl838x_w32_mask(1 << gpios->smi_data, 0, RTL838X_GPIO_PABC_DIR);
+	MDC_WAIT;
+	rtl838x_w32_mask(1 << gpios->smi_clock, 0, RTL838X_GPIO_PABC_DATA);
+	MDC_WAIT;
+	rtl838x_w32_mask(0, 1 << gpios->smi_clock, RTL838X_GPIO_PABC_DATA);
+}
+
+static void smi_write_bits(struct rtl838x_gpios *gpios, u32 data, int len)
+{
+	while (len) {
+		len--;
+		smi_write_bit(gpios, data & (1 << len));
+	}
+}
+
+static void smi_read_bits(struct rtl838x_gpios *gpios, int len, u32 *data)
+{
+	u32 v = 0;
+
+	while (len) {
+		len--;
+		v <<= 1;
+		v |= smi_read_bit(gpios);
+	}
+	*data = v;
+}
+
+/* Bit-banged verson of SMI write access, caller must hold smi_lock */
+int rtl8380_smi_write(struct rtl838x_gpios *gpios, u16 reg, u32 data)
+{
+	u16 bus_id = gpios->bus_id;
+
+	/* Set clock and data pins on RTL838X to output */
+	rtl838x_w32_mask(0, 1 << gpios->smi_clock, RTL838X_GPIO_PABC_DIR);
+	rtl838x_w32_mask(0, 1 << gpios->smi_data, RTL838X_GPIO_PABC_DIR);
+
+	/* Write start bits */
+	smi_write_bits(gpios, 0xffffffff, 32);
+
+	smi_write_bits(gpios, 0x5, 4);		/* ST and write OP */
+
+	smi_write_bits(gpios, bus_id, 5);	/* 5 bits: phy address */
+	smi_write_bits(gpios, reg, 5);		/* 5 bits: register address */
+
+	smi_write_bits(gpios, 0x2, 2);		/* TURNAROUND */
+
+	smi_write_bits(gpios, data, 16);	/* 16 bits: data*/
+
+	smi_z(gpios);
+
+	return 0;
+}
+
+/* Bit-banged verson of SMI read access, caller must hold smi_lock */
+int rtl8380_smi_read(struct rtl838x_gpios *gpios, u16 reg, u32 *data)
+{
+	u16 bus_id = gpios->bus_id;
+
+	/* Set clock and data pins on RTL838X to output */
+	rtl838x_w32_mask(0, 1 << gpios->smi_clock, RTL838X_GPIO_PABC_DIR);
+	rtl838x_w32_mask(0, 1 << gpios->smi_data, RTL838X_GPIO_PABC_DIR);
+
+	/* Write start bits */
+	smi_write_bits(gpios, 0xffffffff, 32);
+
+	smi_write_bits(gpios, 0x6, 4);		/* ST and read OP */
+
+	smi_write_bits(gpios, bus_id, 5);	/* 5 bits: phy address */
+	smi_write_bits(gpios, reg, 5);		/* 5 bits: register address */
+
+	smi_z(gpios);				/* TURNAROUND */
+
+	smi_read_bits(gpios, 16, data);
+	return 0;
+}
+
+static void i2c_pin_set(struct rtl838x_gpios *gpios, int pin, u32 data)
+{
+	u32 v;
+
+	rtl8380_smi_read(gpios, RTL8231_GPIO_DATA(pin), &v);
+	if (!data)
+		v &= ~(1 << (pin % 16));
+	else
+		v |= (1 << (pin % 16));
+	rtl8380_smi_write(gpios, RTL8231_GPIO_DATA(pin), v);
+}
+
+static void i2c_pin_get(struct rtl838x_gpios *gpios, int pin, u32 *data)
+{
+	u32 v;
+
+	rtl8380_smi_read(gpios, RTL8231_GPIO_DATA(pin), &v);
+	if (v & (1 << (pin % 16))) {
+		*data = 1;
+		return;
+	}
+	*data = 0;
+}
+
+static void i2c_pin_dir(struct rtl838x_gpios *gpios, int pin, u16 direction)
+{
+	u32 v;
+
+	rtl8380_smi_read(gpios, RTL8231_GPIO_DIR(pin), &v);
+	if (direction) // Output
```

Comment:
ERROR: do not use C99 // comments
ERROR: trailing statements should be on next line